barter-data-rs/src/dex/tokens/mod.rs

```diff
@@ -0,0 +1,120 @@
+use redis::{Commands, RedisError, Client};
+use core::fmt;
+use std::sync::{Arc, Mutex};
+use lazy_static::lazy_static;
+use serde::{Deserialize, Serialize};
+
+use super::DexError;
+use ethers::{
+  contract::{abigen, ContractError}, 
+  core::types::Address, 
+  providers::{Provider, Http}, 
+  utils::hex::FromHexError
+};
+
+pub struct TokenCache {
+  client: Client,
+}
+
+abigen!(
+  IERC20,
+  r#"[
+    function decimals() external view returns (uint8)
+    function symbol() external view returns (string memory)
+  ]"#,
+);
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Token {
+  pub addr: String,
+  pub symbol: String,
+  pub decimals: u8,
+}
+impl fmt::Display for Token {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+      write!(f, "Token {{ addr: {}, symbol: {}, decimals: {} }}", self.addr, self.symbol, self.decimals)
+  }
+}
+
+lazy_static! {
+  static ref REDIS_CACHE: Mutex<TokenCache> = Mutex::new(TokenCache {
+      client: Client::open("redis://127.0.0.1/").expect("Failed to create Redis client"),
+  });
+}
+
+impl From<RedisError> for DexError {
+  fn from(err: RedisError) -> Self {
```

Comment:
should be able to use the #from macro for all these froms?
same as https://github.com/sector-fi/barter-mono/blob/dev/barter-data-rs/src/error.rs